kip-0019.md

```diff
@@ -0,0 +1,75 @@
+---
+KIP: 0019
+Title: Coin-v6 and fungible-v3
+Author: Robert Soeldner (robert@kadena.io)
+Status: Draft
+Type: Standard
+Category:
+Created: 2023-05-02
+---
+
+## Abstract
+
+The next iteration of the coin contract proposes three main changes to enhance
+the _maintainability_ and _security_ of the contract.
+
+Firstly, the `rotate` function, which is responsible rotating an account guard,
+will be removed to preserve cross-chain soundness. This is to avoid the risk locking funds.
+
+Secondly, the constants and some functions within the contract will be reorganized to
+increase maintainability. This is intended to make the codebase more structured and easier
+to maintain, reducing the chances of errors and improving the overall quality of the contract.
+
+Finally, verification related code of the coin contract will be modified to allow for
+formal verification. This will enable rigorous proves of the contract's code to ensure
+its correctness and reliability. By incorporating these changes, the new iteration of the
+coin contract aims to provide a more secure and efficient platform for the exchange of
+cryptocurrencies.
+
+## Motivation
+
+The presence of the `rotate` function in the current implementation of the contract undermines the
```

Comment:
I highly recommend not to remove `rotate` as this is one of our unique selling points over other chains where the account/address is directly tied to the public/private keypair. If your private key get's compromised on such chain you are forced to move all existing assets to a new address(keypair).

Kadena's chain has the unique property that our address is not tied to the keypair (or keyset in our case). So being able to rotate is a huge benefit. Allowing users to keep the same address, but changing their keyset to sign with.

Reply:
I agree with @EnoF here that _rotation_ is unique to the Pact language in that you can rotate the underlying governance of a particular account in the coin contract, but there is a distinction here that needs to be made: 

The removal of `rotate` will *not* remove the ability for accounts to have rotatable governance. This is a misconception. It *will* remove the ability to *arbitrarily* rotate account governance from any guard to any guard. However, the feature still remains that if one uses a keyset reference and stores a `keyset-ref-guard`, then keys are rotateable via `define-keyset`. Since keysets are namespaced, users may unambiguously name and store guards they expect to rotate, and the principal account type associated with them is the `r:` protocol, so they may be stored under unique account addresses as well.
  
So the question here that needs to be answered is this: should guards be arbitrarily rotatable, or should rotation be restricted to when it makes sense? `rotate` makes any crooss-chain transaction possibly unsound in its presence: funds can be locked up if any call to `rotate` occurs in between the burn and creation steps on the two chains. That makes for a terrible UX! I don't think restricting rotation to just builtin Pact primitives would create the same burden of UX because there already exists a recourse for users that care about rotation.

Reply:
As an addendum: also, keep in mind that `rotate` *breaks the invariants for principal account protocols*. This is extremely important! While [`enforce-reserved`](https://github.com/kadena-io/chainweb-node/blob/12c2c42431200db5fc33ffe5ae859d756c93ae31/pact/coin-contract/v5/coin-v5.pact#L509) in the coin contract requires that `k:` accounts be well-formed when doing any useful operation in the coin contract, one can still call `rotate` on principal accounts, effectively *bricking* them until they are rotated back to the correct governance defined by the principal. There are only negatives here: principals don't work as intended, *rotate* doesn't provide any value, *and* we're going to be restricting `create-account` so that users can only interact with `coin` if they have a principal account protocol in play. The existence of `rotate` in this case only introduces surface area for griefing, with no benefits.

Reply:
If I understood this correctly, this means that users should be encouraged to create accounts like:
```
(coin.create-account "k:mykey" (keyset-ref-guard "r:my-keyset"))
```
And then they will be able to rotate said keyset like:
```
(define-keyset "r:my-keyset" (read-keyset 'new-keyset))
```
Which then makes the `coin.transfer-crosschain` reference the defined keyset rather than the value of the keyset and therefore you get the SPV to become sound? If this is correct we should probably some more "encouraging documentation" around this for new devs. At least to me the use of `define-keyset` did not ring a bell to be used in this way.

Reply:
@EnoF close! You understood correctly, but for the interest of the KIP process, there are some details to iron out. Namely, let's look at the call you've written:

```lisp
 (coin.create-account "k:mykey" (keyset-ref-guard "r:my-keyset"))
```

This would not be correct because the account *name* would be `r:<my-keyset>`, while the keyset ref guard would be called to wrap around a particular predefined keyset. In fact, the call to creating a rotatable account would be something like this in general: 

```lisp
(coin.create-account (create-principal (keyset-ref-guard "<some keyset name>")) (keyset-ref-guard "<some keyset name>"))
```

The call to `create-principal` creates an `r:`-principal account name much in the same sense as `k:`. They look like `r:<some keyset name>`. And yes, the underlying keyset is fully rotateable via calls like: 

```lisp
(define-keyset "<some keyset name>" (read-keyset 'new-keyset))
```


> Which then makes the coin.transfer-crosschain reference the defined keyset rather than the value of the keyset and therefore you get the SPV to become sound? If this is correct we should probably some more "encouraging documentation" around this for new devs. At least to me the use of define-keyset did not ring a bell to be used in this way.

Yes, exactly. If the keyset is defined on the target chain, then the keys may rotate underneath it, but the keyset name *reference* is a static string, which does not change. We have documentation for this in the principal account section of the readthedocs, but I'll work on writing up a concrete tutorial/introduction and tie it in with this proposal to ease everyone's minds.

Reply:
@emilypi I tried creating an account using a `keyset-ref-guard`, 
which requires me to use `define-keyset`. I was hoping to define it like:

``` 
(define-keyset
  (create-principal (read-keyset 'keyset))
  (read-keyset 'keyset))
```

In the hopes that the keyset would be guarded like a `k:account` would be.
This is not possible because you need to be on a namespace to use `define-keyset`.
Which means that I would use either `free` or `user`, my keyset will not be guarded.

Then I thought maybe I could use `ns-create-principal-namespace` to solve this problem:

```
(define-namespace
  (ns.create-principal-namespace (read-keyset 'keyset))
  (read-keyset 'keyset)
  (read-keyset 'keyset))

(namespace (ns.create-principal-namespace (read-keyset 'keyset)))

(define-keyset
  (format "{}.{}" [(ns.create-principal-namespace (read-keyset 'keyset )) "keyset"])
  (read-keyset 'keyset))

(coin.create-account 
  (create-principal 
    (keyset-ref-guard 
      (format 
        "{}.{}" 
        [(ns.create-principal-namespace (read-keyset 'keyset )) "keyset"])))
    (keyset-ref-guard 
      (format 
        "{}.{}" 
        [(ns.create-principal-namespace (read-keyset 'keyset )) "keyset"])))
```

This should prevent your keyset ref from being squatted, but it is quite a bit of boilerplate.

I think it would be a good compromise between UX/DX and security if we would
upgrade the `create-account` to auto define a keyset in a way that only the original
keyset can sign for. (i.e. `k:pubkey` the `pubkey` needs to sign for the initial
`define-keyset`). Maybe this wouldn't work on the `define-keyset`, but we could maybe
introduce something like `ks.create-principal-keyset` that would do this for you.

Reply:
@EnoF I see your point using `create-principal-namespace` and I think `define-keyset` could implicitly use the current namespace. Maybe there are good reasons for explicitly writing the NS @emilypi ?

For later reference, repl script using the KIP-0019 infrastructure showing key-rotation.
```
(begin-tx)
(env-data {"robert": {"keys": ["key1"], "pred": "keys-any"}})
(env-sigs [{'key: "key1", 'caps: []}])

(define-namespace 'free (read-keyset "robert") (read-keyset "robert"))
(namespace 'free)
(define-keyset "free.robert" (read-keyset "robert"))
(commit-tx)


(begin-tx)
(namespace 'free)
(env-data {"andy": {"keys": ["key2"], "pred": "keys-any"}})
(env-sigs [{'key: "key2", 'caps: []}])

(expect-failure
 "cant rotate keys due to missing keyset"
 (define-keyset "free.robert" (read-keyset "andy")))
(commit-tx)

(begin-tx)
(namespace 'free)
(env-data {"robert": {"keys": ["key1"], "pred": "keys-any"},
             "andy": {"keys": ["key2"], "pred": "keys-any"}})
(env-sigs [{'key: "key2", 'caps: []} {'key: "key1", 'caps: []}])

(define-keyset "free.robert" (read-keyset "andy"))
(commit-tx)


(begin-tx)
(namespace 'free)
(load "fungible-xchain-v1.pact")
(load "fungible-v3.pact")
(load "coin-v6.pact")


(create-table coin.coin-table)
(create-table coin.allocation-table)

(coin.create-account
  (create-principal (keyset-ref-guard "free.robert"))
  (keyset-ref-guard "free.robert"))
(commit-tx)
```

Reply:
Note that `free.robert` can be squated by anyone on a different chain. So you could have defined `free.robert` with your keyset on chain 1, while I would define `free.robert` on chain 2 with a different keyset. So to prevent this from happening, we need to have `define-keyset` enforce the initial keyset used to create the definition, i.e. `k:pubkey` or a different principal that could enforce such requirement.